File: src/main/java/com/numberbox/common/handler/GlobalExceptionHandler.java
//package com.numberbox.common.handler;
//
//import lombok.val;
//import org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController;
//import org.springframework.http.HttpStatus;
//import org.springframework.http.ResponseEntity;
//import org.springframework.web.bind.MethodArgumentNotValidException;
//import org.springframework.web.bind.MissingServletRequestParameterException;
//import org.springframework.web.bind.annotation.ExceptionHandler;
//import org.springframework.web.bind.annotation.RestControllerAdvice;
//import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
//
//@RestControllerAdvice
//public class GlobalExceptionHandler {
//    @ExceptionHandler(value = MethodArgumentNotValidException.class)
//    public ResponseEntity<Object> handlingModelValidException(MethodArgumentNotValidException ex): ResponseEntity<CustomResTmpl<ExceptionMsgTmpl>> {
//        BasicErrorController
//        // 클라이언트에 전달할 메시지(@Valid에 적용된 message)
//        var clientMsg = ex.bindingResult.fieldError?.defaultMessage ?: "입력양식이 올바르지 않습니다."
//        if (ex.bindingResult.fieldError?.isBindingFailure == true) clientMsg = "요청 형식이 올바르지 않습니다."
//        // 예외 메시지 템플릿 작성
//        val exCode = ExceptionCode.INVALID_OBJECT_PROPERTY
//        val exInfo = ExceptionMsgTmpl.makeExceptionMsg(exCode)
//        // 예외 로그 찍기
//        println("errInfo : $exInfo \nerrDesc : $ex")
//        // 응답 객체 전달
//        val resBody = CustomResTmpl<ExceptionMsgTmpl>(showMsgAlert = true, message = clientMsg, data = exInfo)
//        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(resBody)
//    }
//
//    @ExceptionHandler(value = [MissingServletRequestParameterException::class])
//    fun handlingMissingRequestParameterException(ex:MissingServletRequestParameterException): ResponseEntity<CustomResTmpl<ExceptionMsgTmpl>> {
//        // 클라이언트에 전달할 메시지(@Valid에 적용된 message)
//        val clientMsg = "유효한 요청이 아닙니다."
//        // 예외 메시지 템플릿 작성
//        val exCode = ExceptionCode.MISSING_REQUEST_PARAM
//        val exInfo = ExceptionMsgTmpl.makeExceptionMsg(exCode)
//        // 예외 로그 찍기
//        println("errInfo : $exInfo \nerrDesc : $ex")
//        // 응답 객체 전달
//        val resBody = CustomResTmpl<ExceptionMsgTmpl>(showMsgAlert = true, message = clientMsg, data = exInfo)
//        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(resBody)
//    }
//
//    @ExceptionHandler(value = [MethodArgumentTypeMismatchException::class])
//    fun handlingMethodArgumentTypeMismatchException(ex:MethodArgumentTypeMismatchException): ResponseEntity<CustomResTmpl<ExceptionMsgTmpl>> {
//        // 클라이언트에 전달할 메시지(@Valid에 적용된 message)
//        val clientMsg = "유효한 요청이 아닙니다."
//        // 예외 메시지 템플릿 작성
//        val exCode = ExceptionCode.PARAM_TYPE_MISS_MATCH
//        val exInfo = ExceptionMsgTmpl.makeExceptionMsg(exCode)
//        // 예외 로그 찍기
//        println("errInfo : $exInfo \nerrDesc : $ex")
//        // 응답 객체 전달
//        val resBody = CustomResTmpl<ExceptionMsgTmpl>(showMsgAlert = true, message = clientMsg, data = exInfo)
//        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(resBody)
//    }
//
//    @ExceptionHandler(value = [BusinessRuleInvalidException::class])
//    fun handlingBusinessException(ex: BusinessRuleInvalidException): ResponseEntity<CustomResTmpl<ExceptionMsgTmpl>> {
//        // 클라이언트에 전달할 메시지(@Valid에 적용된 message)
//        val clientMsg = ex.msg
//        // 예외 메시지 템플릿 작성
//        val exInfo = ExceptionMsgTmpl.makeExceptionMsg(ExceptionCode.BUSINESS_RULE_INVALID)
//        // 예외 로그 찍기
//        println("errInfo : $exInfo \nerrDesc : $ex")
//        // 응답 객체 전달
//        val resBody = CustomResTmpl<ExceptionMsgTmpl>(showMsgAlert = false, message = clientMsg, data = exInfo)
//        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(resBody)
//    }
//}

File: project/user-interface/rest-api/src/main/kotlin/com/kamcci/numberbox/restapi/handler/exception/GlobalExceptionHandler.kt
package com.kamcci.numberbox.restapi.handler.exception

import com.kamcci.numberbox.app.domain.exception.BusinessValidException
import com.kamcci.numberbox.restapi.util.response.ResponseUtil
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

/**
 * 컨트롤러 예외 처리 목적 aop 프록시
 */
@RestControllerAdvice
class GlobalExceptionHandler : ResponseEntityExceptionHandler() {

    // MethodArgumentNotValid
    override fun handleMethodArgumentNotValid(
        ex: MethodArgumentNotValidException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {
        logger.warn(ex)
        return ResponseUtil.error(ex, status, request)
    }

    @ExceptionHandler(value = [BusinessValidException::class])
    fun handleBusinessInValidException(
        ex: Exception,
        body: Any?,
        request: WebRequest
    ): ResponseEntity<Any> {
        println(ex.stackTraceToString())
//        logger.warn(ex)
        return ResponseUtil.error(ex, HttpStatus.BAD_REQUEST, request)
    }

    @ExceptionHandler(value = [Exception::class])
    fun handleException(
        ex: Exception,
        body: Any?,
        request: WebRequest
    ): ResponseEntity<Any> {
        logger.error(ex)
        return ResponseUtil.error(ex, HttpStatus.INTERNAL_SERVER_ERROR, request)
    }


    /**
     * 모든 예외타입에 대하여 응답 형식 지정
     */
    override fun handleExceptionInternal(
        ex: Exception,
        body: Any?,
        headers: HttpHeaders,
        statusCode: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any> {
        logger.warn(ex)
        return ResponseUtil.error(ex, statusCode, request)
    }
}